operator/pkg/nodepools/pools.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package nodepools exists currently because it is required to be a separate due to import conflicts.
+// Since nearly every package depends on vectorizedv1alpha1, we have to move
+// some functions out of it, because vectorizedv1alpha1 can not import anything
+// that depends on it (and since almost everything depends on
+// vectorizedv1alpha1, it can basically import almost nothing)
+//
+// Practically, vectorizedv1alpha1 -> pkg/labels is impossible, so we move these functions out of vectorizedv1alpha1.
+package nodepools
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	vectorizedv1alpha1 "github.com/redpanda-data/redpanda-operator/operator/api/vectorized/v1alpha1"
+	"github.com/redpanda-data/redpanda-operator/operator/pkg/labels"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetNodePools gets all NodePoolSpecs within a cluster.
+// This also includes deleted node pools. These are removed from spec,
+// therefore a NodePoolSpec is synthesized by searching for existing StatefulSets.
+// To include information that a NodePool has been deleted, the NodePoolSpec is
+// wrapped into a type with an extra Deleted boolean.
+func GetNodePools(ctx context.Context, cluster *vectorizedv1alpha1.Cluster, k8sClient client.Reader) ([]*vectorizedv1alpha1.NodePoolSpecWithDeleted, error) {
+	var nodePoolsWithDeleted []*vectorizedv1alpha1.NodePoolSpecWithDeleted
+
+	nps := cluster.GetNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
+
+		nodePoolsWithDeleted = append(nodePoolsWithDeleted, &vectorizedv1alpha1.NodePoolSpecWithDeleted{
+			NodePoolSpec: np,
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(ctx, &stsList, &client.ListOptions{
+		LabelSelector: labels.ForCluster(cluster).AsClientSelector(),
+	})
+	if err != nil {
+		return nil, err
+	}
+outer:
+	for i := range stsList.Items {
+		sts := stsList.Items[i]
+
+		for _, ownerRef := range sts.OwnerReferences {
```

Comment:
I find this check and the later check for the existence of a `redpanda` container a bit odd. We're already using a label selector to filter our list of Sts's; is there a reason we're performing these extra checks? Are they checking for different things? If they're not, why does one error but the other does not?

I like to avoid "anxious" code when possible as it creates an ethical dilemma  when it comes time to refactor. I'll have all the above questions with potentially no one to answer them nor test cases to validate them.

If these checks are in place for a reason, document what exactly they're checking for and ideally the situations in which they might trigger.

If these checks are just anxiety or paranoia that you can't shake, group them together and label them as such. A simple `// Paranoid check to make sure we don't delete other STSs` can save so much head-aching in the future.